### DISCUSSION TITLE
Prevent new node installs from downgrading packages

### DISCUSSION THREAD

**Discussion by adin00** [Comfy-Org/ComfyUI-Manager#980] [Ideas]
I'm not even sure if this can be done, I'm just throwing this out there. 

First idea that comes to mind is a dry-run of sorts, and warn about any downgrade, with maybe a acontinue yes/no modal, and I would hapily skip that node until it is updated, or at all.
As it is at this rn, I can either check out the requirements of the repo and compare with current installation (very tedious) or just install it, see if/what it breaks, and take it from there. 

Another idea is the ability to roll-back a faulty install. I've noticed that a backup of current nodes list is created before any node install, but I have no idea how to use that. furthermore, if the node dependencies prevents comfyui from running entirely, I'm pretty sure that json can't be used in any way.

I know that the same thing can be said about upgrading packages, but from my experience this case might break some other custom nodes, but never the host, which is much better from my perspective.

**Comment by ltdrdata** [comment]
1. I have also considered the yes/no modal option, but I am still deliberating on it. The issue is that I need to find a solution for cases where multiple packages are being installed, as this could lead to the modal appearing very frequently during a single installation process.

2. At present, the backup/restore functionality is limited as it only operates at the version level of custom nodes. I am considering implementing it at the pip package level. 
However, I am somewhat hesitant about this approach because it would require very extensive installation changes. 
I'm thinking of applying this in a way that allows it to be selectively applied during the restore dialog.

**Reply by bigcat88** [reply]
Is there any functionality that could be declared via config or something so that some packages would not be downgraded? (for example, to put their pinned version at the end of the update/reinstall actions)

I am rewriting my project here so that it would start using Manager - since you easily accept PRs (for which I thank you very much!) - it is easier to fix everything in your repository than to repeat this functionality.
And I ran into a situation where one node is dependent on the `inference-gpu` package - and that package strictly specifies the old version of FastAPI.
As a result, after the update, the entire project does not work at all, because it is not compatible with the old version of FastAPI.

If it were possible to somehow define an action at the end of any update or installation, so that at the end the required python package would always be of the required version - it would be good.

I just haven't figured out yet how to do this correctly

**Reply by bigcat88** [reply]
e.g. ability to define custom actions somewhere like here:

https://github.com/ltdrdata/ComfyUI-Manager/blob/e8111527b48291cde5a27d69143d91aa3d01d6d6/glob/manager_util.py#L309

will make life much easier.

What do you think about this?

**Reply by ltdrdata** [reply]
Now you can use `pip_auto_fix.list`.

https://github.com/ltdrdata/ComfyUI-Manager#paths

```
wget>=3.2
yarl!=1.18.2
tyro>=0.8.5
```

**Comment by ghost** [comment]
Bumping to say I really would like this feature. In fact, the default condition should be to disallow downgrading IMO. It only makes sense, since nodes do tend to become abandoned, and they don't update their requirements.  I just installed a custom node that downgraded my torch and broke everything. Very uncool.

**Comment by audovoice** [comment]
I could maybe be an option something like "debug install". 
people sometimes know when a difficult node will be an issue and want to risk it anyway, only for it to blow up their whole installation.

 I found this 
```
Overview:
    Install new packages without upgrading installed ones: pip install (without -U)
    Upgrade only packages that are outdated according to requirements: pip install --upgrade --upgrade-strategy only-if-needed (default in new versions)
    Upgrade package and all dependencies to latest version: pip install --upgrade --upgrade-strategy eager (default in old versions)
    Install or upgrade listed packages without touching dependencies: --no-deps
```
But I am not sure that is a feature on the version of pip comfyui uses. 
https://stackoverflow.com/questions/2861183/upgrade-package-without-upgrading-dependencies-using-pip

To be honest even something like a "manual" setting to just ignore trying to resolve a requirements.txt and let the user try to fix it themselves.

**Reply by MavenDE** [reply]
This is super important feature to have. NEVER downgrade any packages, unless a consent from the user is given.
